### PR TITLE
8287071: Loom: Cleanup x86_64 gen_continuation_enter

### PR DESCRIPTION
**Pull request by shipilev**
During x86_32 porting, I ended up rewriting the copy of x86_64 gen_continuation_enter significantly to make it more streamlined. The original x86_64 gen_continuation_enter should enjoy the similar streamlining, since it would likely to be the template for many other architectures.

There might be more things to do, I would not know until x86_32 fully works. I am pretty sure current thing is self-sufficient.

Brief tour of changes:
 - Check the incoming registers explicitly
 - Pass explicit argument registers to `fill_continuation_entry`
 - Use the usual `NativeCall` alignment code to produce a patchable call
 - Check that static stub is emitted, fail hard otherwise
 - Handle register shuffling for `exception_handler_for_return_address` the similar way we do it in `StubGenerator`

Additional testing:
 - [x] Linux x86_64 fastdebug, prospective `jdk_loom hotspot_loom`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287071](https://bugs.openjdk.org/browse/JDK-8287071): Loom: Cleanup x86_64 gen_continuation_enter

### Reviewers
 * [Ron Pressler](https://openjdk.java.net/census#rpressler) (@pron - Author) ⚠️ Review applies to [3c721970](https://git.openjdk.java.net/jdk/pull/8802/files/3c7219708e9ffcd384310d8cdfc5cdd1937c4ec3)
 * [Patricio Chilano Mateo](https://openjdk.java.net/census#pchilanomate) (@pchilano - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8802/head:pull/8802` \
`$ git checkout pull/8802`

Update a local copy of the PR: \
`$ git checkout pull/8802` \
`$ git pull https://git.openjdk.java.net/jdk pull/8802/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8802`

View PR using the GUI difftool: \
`$ git pr show -t 8802`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8802.diff">https://git.openjdk.java.net/jdk/pull/8802.diff</a>

</details>

### CODE REVIEW

**Comment by shipilev** [comment]
Still open for review. Windows test failures are known and handled separately.

**Comment by shipilev** [comment]
Still open for reviews.

**Comment by shipilev** [comment]
Still open for reviews, with RDP1 looming. Please review!

**Review by pron — APPROVED** [review]
Looks fine. Thanks.

File: src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp [hunk 1]
```diff
@@ -1259,99 +1259,129 @@ static void verify_oop_args(MacroAssembler* masm,
...
+                                              const char* name) {
+  assert(!actual_vmreg->is_stack(), "%s cannot be on stack", name);
+  assert(actual_vmreg->as_Register() == expected_reg,
+         "%s is in unexpected register: %s instead of %s",
+         name, actual_vmreg->as_Register()->name(), expected_reg->name());
+}
+
 static void gen_continuation_enter(MacroAssembler* masm,
                                  const methodHandle& method,
                                  const BasicType* sig_bt,
```
Comment (pchilano) [inline src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:1276]: Should we remove these two parameters? They were used on a commented out line which is now removed.

**Review by pchilano — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]
Looks good. Do you plan similar change for Aarch64 code?

**Comment by shipilev** [comment]
> Looks good. Do you plan similar change for Aarch64 code?

Yes, but not in JDK 19 timeframe. I want x86_64 code to be as clean as possible to serve as template for porting to other arches.

**Comment by vnkozlov** [comment]
> > Looks good. Do you plan similar change for Aarch64 code?
> 
> Yes, but not in JDK 19 timeframe. I want x86_64 code to be as clean as possible to serve as template for porting to other arches.

Okay.

**Pull request closed without merging**